### PR TITLE
Improved rustdoc rendering for unstable features

### PR DESCRIPTION
**Pull request by Manishearth**
This replaces "unstable" with "this is an experimental API", and uses a `<details>` tag to expand to the reason.

The `<details>` tag renders as a regular div (with the details show) on browsers which don't support it, On browsers which do support it, it shows only the summary line with an expandy-arrow next to it, and on clicking it the details will turn up below it.

This is somewhat a strawman proposal. The main issue is that we need to improve our messaging around unstable APIs. Since they turn up in the docs, we should be clearer that they are experimental (and perhaps add something about nightly-only). I'm making this PR to kickstart discussion on this.

Example rendering: http://manishearth.github.io/rust-internals-docs/std/io/trait.Read.html#method.chars

screen shot 2017-01-04 at 10 15 37 pm

expands to

screen shot 2017-01-04 at 10 15 43 pm

cc @steveklabnik @jdub

### CODE REVIEW

**Comment by Manishearth** [comment]
"This is a nightly-only experimental API" might be better.

Changing the text (or even the fact that it is displayed) based on the version of the docs being built is also interesting.

**Comment by nagisa** [comment]
> "This is a nightly-only experimental API" might be better.

“Unstable API”?

---

I like the use of `<summary>`+`<detail>`. We could use that for actual documentation as well instead of using the javascript junk to emulate expandable stuff as an improvement later on as well.

**Comment by Manishearth** [comment]
Part of the point was to remove the word "unstable" as it can cause confusion.

**Comment by jdub** [comment]
Nice! I just saw your suggestion to open an RFC… this seems more straightforward. 😄 

It's getting a bit wordy, but perhaps: "⚠️  This is an experimental API, available only on Rust nightly builds". It may seem superfluous to say "Rust", but it helps clarify the "nightly" jargon.

**Comment by GuillaumeGomez** [comment]
"unstable API" seems fine for me, but otherwise I don't much preference in here.

Big 👍 for me. This is a huge improvement and really looking forward to it.

**Comment by aturon** [comment]
@Manishearth This is a great change. I agree that moving away from the stable/unstable terminology will reduce confusion. Adding "nightly-only" is probably a good idea, for maximum clarity.

cc @brson

**Comment by nikomatsakis** [comment]
I too prefer experimental to unstable. It makes us sound like scientists. I would prefer to add some mention of nightly as well.

/me wants a lab-coat

**Comment by Manishearth** [comment]
I'm disappointed that there is no Erlenmeyer flask emoji. Microscope (🔬) exists, and it is supposed to combine with person emojis to create scientist emojis, but that doesn't work on many fonts right now.

So something like:

 - "⚠️ This is an experimental API, available only on Rust nightly builds"
 - "🔬 This is a nightly-only experimental API"

is probably what we're settling on.

**Comment by rkruppe** [comment]
FWIW, the ⚠️ emoji is already used to mark unsafe functions.

**Comment by Manishearth** [comment]
I don't see that being used in the docs anywhere?


I noticed that not all APIs have extended unstable reasons, so I should probably fix up this PR to handle that (e.g. http://manishearth.github.io/rust-internals-docs/syntax/parse/fn.raw_str_lit.html)

**Comment by steveklabnik** [comment]
https://github.com/rust-lang/rust/pull/37250

**Comment by jdub** [comment]
@Manishearth Microscope is way cooler. 😄

**Comment by liigo** [comment]
🍼 Baby Bottle (U+1F37C)

🍼 This is an experimental API, available only on Rust nightly builds.
or just the icon 🍼 with a popup tooltip.

**Comment by Manishearth** [comment]
screen shot 2017-01-06 at 10 16 32 am

**Comment by nikomatsakis** [comment]
@Manishearth I can't really make out that it is a microscope. :( I sort of like baby bottle. =)

**Comment by Manishearth** [comment]
Not really better with the baby bottle

screen shot 2017-01-06 at 5 18 42 pm


Note that emoji will render differently on different platforms. What we should do here is just make the font size a bit larger.

**Comment by Manishearth** [comment]
How does a font size of 1.5em sound?

screen shot 2017-01-06 at 5 20 10 pm

**Comment by GuillaumeGomez** [comment]
@nikomatsakis: And I prefer the microscope. Should we vote? :p

**Comment by Manishearth** [comment]
(I'll need to update the tests too, so don't directly r+ this, just mention "r=me" or whatever. I won't update them now since we have to yet settle on a solution)

**Comment by liigo** [comment]
> Note that emoji will render differently on different platforms.

There is EmojiOne. image image

**Comment by nikomatsakis** [comment]
Microscope is fine with me. The bigger size looks mildly better.

**Comment by Manishearth** [comment]
Are we still in the bikeshed phase for this PR or should I start cleaning up the tests so that we can land this?

**Comment by GuillaumeGomez** [comment]
From my point of view, we're in the details now. Text shouldn't change and it seems microscope is the preferred icon. Maybe wait confirmation from others and then clean up?

**Comment by aturon** [comment]
@Manishearth let's ship! Agreed on larger-sized microscope.

**Comment by Manishearth** [comment]
Updated tests.

**Comment by GuillaumeGomez** [comment]
Then it's all good. Thanks!

@bors: r+

**Comment by alexcrichton** [comment]
@bors: retry

* hang in compiling librustc?
